### DISCUSSION TITLE
Cordova Plugin File doesn't work on Android 13

### DISCUSSION THREAD

**Discussion by leonelgregioauler** [apache/cordova#424] [Q&A] [answered]
Good morning! 

What permission should I add to be able to write a .CSV file in the "Documents" folder ? 

I've tried all the permissions below. 

![image](https://github.com/apache/cordova/assets/12154804/e51043b9-e57d-4353-a53d-8320115d9a31)

I got this error: "**onErrorGetDir**". The folder is: **file:///storage/emulated/0/**

![image](https://github.com/apache/cordova/assets/12154804/96fb28cc-aa88-436e-8124-c8de11edffb6)

Remember, this works fine on others versions, like Android 11 or below. Android 12 I haven't tested yet, but it seems to work !

Thank you for now !

**Comment by breautek** [comment, marked as answer]
Android's external file system is now quite limiting and doesn't allow direct file access via file APIs for all files anymore. It does allow for media files such as images, videos, audio. For other file types, the native MediaStore API needs to be used, which the file plugin does not use (and it's not really a file API so it's going to be hard to facade it to make it look like a file API).

> Remember, this works fine on others versions, like Android 11 or below. Android 12 I haven't tested yet, but it seems to work !

This wouldn't have worked since Scoped Storage, which was introduced in API 29, but existing apps could opt out and revert back to legacy filesystem. Starting API 30, the scoped storage is fully enforced.

The short answer, if you need external storage access, you'd be probably better off using a plugin that actually interfaces with the native MediaStore API such as [cordova-plugin-saf-mediastore](https://www.npmjs.com/package/cordova-plugin-saf-mediastore) (not an endorsement, never used it myself, and you may want to check the license cause it's not a license I'm familiar with).

The filesystem works well with internal storage (e.g. apps own data directory). I'm not sure if the filesystem plugin can be updated to use the MediaStore API while keeping it's old API signature, so support for external storage may just drop entirely altogether. These are my own thoughts however, and this hasn't been brought up for a community discussion yet on the mailing list.

**Reply by ghost** [reply]
Be that as it may... am having issues making my app work with the new plugin. When I take a picture using the camera plugin it returns a fileURI  which I was always able to send to the Getfile() method. This no longer works. I have been banging my head trying to figure out to get a photo from the cacheDiirectory and move it to my dataDirectory/Photos folder (renaming it along the way). The worked fine before all thse file system changes. I tried using window.resolveLocalFileSystemURL(), to URL, messing with path strings. Can anyone one help??

**Reply by breautek** [reply]
In short, the filesystem no longer allows you to modify any files not owned by your app. File modifications can only be done through a media store API.

If you want to move the file... you wouldn't be able to actually move it (cause that is considered a write, a delete of the original file) but you should be able to **copy** it.

**Reply by ghost** [reply]
The photo plugin returns  "fileURI " =  file:///data/user/0/com.nsbasic.Vantastic/cache/1697768037972.jpg
(inside the cache folder of the apps sandbox). Simply passing the above file URI to the getfile() method returns a not found error. How can i convert the above "fileURI"  to something that the getfile() can find? I have tried:
1. window.resolveLocalFileSystemURL(fileURI) 
2. window.resolveLocalFileSystemURL(cordova.file.cacheDirectory) + "1697768037972.jpg"
3. toURL(fileURI)
Any ideas??

**Reply by ghost** [reply]
This code used to work:

//---Begin Photo Stuff
Mimage.onclick = function() {
    options = {
        quality: 75,
        correctOrientation: true,
        allowEdit: false,
        destinationType: 1
    };
    navigator.camera.getPicture(onPictureReturned, onPictureError, options);
};
function onPictureReturned(fileURI) {
    varLastSlash = InStrR(fileURI, "/", 0);
    varfileURI = fileURI;
    console.log(varfileURI);
    PhotoName = Mid(fileURI, varLastSlash); //The Photo name
    varPicture = Trim(gSelectedBroker) + "-" + Trim(gSelectedCategoryNumber) + "-" + Trim(gSelectedChainNumber) + "-" + Trim(gSelectedStoreNumber) + "-" + CStr(Year(varDate)) + PhotoMonth + PhotoDay + "-" + Trim(varQuestionNO);
    varPicture = varPicture + ".jpg";
    MovePicS();
}
function onPictureError(message) {
    NSB.MsgBox("Failed because " + message);
}
function MovePicS() {
    gfileSystem.root.getFile(varfileURI, {
        create: false
    }, gotFileEntryS, fail2st);
}
function gotFileEntryS(fe) {
    FileEntry = fe;
    FileEntry.moveTo(gdataDir, varPicture, SubSuccess2, fail3st);
}
function SubSuccess2(entry) {
    //MsgBox "Successfully Moved file from Cache:" & vbCRLF & PhotoName & vbCRLF & "To Photos: " & varPicture
    NSB.MsgBox("Press [SAVE ANSWER] to Complete Process...", 0, "Photo Successfull");
}
function fail2st(error) {
    console.log(error);
    NSB.MsgBox("Error getFile(PhotoName: " + error.code + " -" + error.error + "-" + PhotoName);
}
function fail3st(error) {
    NSB.MsgBox("Error .moveTo(gdataDir: " + error.code + " -- " + error.error + "-" + varPicture);
}
//---End Photo Stuff

**Reply by mvandak** [reply]
We had the same problem with writing to our own directory in Android root dir starting with Android 13.

As far as I understood from the Android documentation, `MANAGE_EXTERNAL_STORAGE` is enough to write to that directory. This permission was implemented in one of previous Android versions. However, this permission has to be granted exclusively by user, it is not enough to be mentioned in `config.xml`.
We managed it in our application by checking permission at app start and if not granted, app starts intent to request permission to be granted by user. This we had done already some time ago in previous version of Android.

But, starting with Android 13 we were not able to write. I realized that the cordova-file-plugin is refusing get dir request because we had have not granted all `READ_MEDIA_*` permissions and`FileUtils.hasReadPermission()` returned false. If I added `READ_MEDIA_*` permissions to `config.xml` everything worked as before.

Anyway, I don't understand why filePlugin is testing if all `READ_MEDIA_*` permissions are granted to allow read access and not at least one of. IMHO, this is bypassing the Android granularity.
And also `MANAGE_EXTERNAL_STORAGE` should be taken into account when testing read/write permissions.

I propose to test in `hasReadPermission` and also in `hasWritePermission` if all files access is granted also. If it is granted, everything works without `READ_MEDIA_*` permissions granted. In case, I disabled `All files access` in Android Special access settings, read/write was refused with `Unknown error` from filePlugin
I propose these methods to be modified in `FileUtils` as follows:

```java
 private boolean hasReadPermission() {
        if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            return Environment.isExternalStorageManager() // Test all files access, MANAGE_EXTERNAL_STORAGE permission
                    || PermissionHelper.hasPermission(this, Manifest.permission.READ_MEDIA_IMAGES)
                    || PermissionHelper.hasPermission(this, Manifest.permission.READ_MEDIA_VIDEO)
                    || PermissionHelper.hasPermission(this, Manifest.permission.READ_MEDIA_AUDIO);
        } else {
            return PermissionHelper.hasPermission(this, Manifest.permission.READ_EXTERNAL_STORAGE);
        }
    }

private boolean hasWritePermission() {
    // Starting with API 33, requesting WRITE_EXTERNAL_STORAGE is an auto permission rejection
    return android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU
            ? Environment.isExternalStorageManager()   // Test all file access, MANAGE_EXTERNAL_STORAGE permission
            : PermissionHelper.hasPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE);
}
```

**Comment by leonelgregioauler** [comment]
Do you have any tips for opening a file browser so that the user can choose where to save the content ? The content would be generated by the app.

**Reply by breautek** [reply]
This isn't something I do in my own apps so I don't have first hand experience but I'm pretty sure file/folder picking is facilitated by the Android's MediaStore APIs. Generally using the API requires using intents which opens the system file browser for those kinds of actions.

The aforementioned plugin for example has a `selectFolder` api in which it "Launches an Intent to select a folder to which files can be saved.".

Android's scoped storage mechanism is privacy-focused so programmatic control is quite limited.

**Reply by leonelgregioauler** [reply]
Yes, I understand !

I'll try to use the mentioned plugin. 

Thanks a lot for the tips !!

**Reply by pixellet14** [reply]
@leonelgregioauler hi, have you been able to create a folder using that plugin yet? I am trying to create a folder inside the DCIM directory for my app and then save the user modified pictures in that but unable to do so. Please can you help me? I shall be grateful. Thanks in advance

**Reply by leonelgregioauler** [reply]
Hi @pixellet14 !! 

In my implementation, I used "writeFile" function. It have one parameter "folder" that I couldn't use it and one parameter "subFolder" that works fine !!

1 - The "folder" parameter doesn't make difference, that is, it does nothing. I have doubt if I don't know how to use or if this plugin have a bug

2 - The "subFolder" parameter write a file into "Downloads" and "Documents" folder at same time. The subfolder informed into this parameter is created and the file stays inside it.

3 - You can see my implementation, on my GitHub project. See files "httpUtil.js" and "dashboard-mensal.js"

    https://github.com/leonelgregioauler/App-Contador-Pessoas-CP-3001/blob/main/src/js/httpUtil.js 
    https://github.com/leonelgregioauler/App-Contador-Pessoas-CP-3001/blob/main/src/js/viewModels/dashboard-mensal.js

4 - Try saving a single "txt" file or "csv" file to make your test easier.

If this is not good for you, you can try using "saveFile" function !! 

Tell me more later !!

**Reply by pixellet14** [reply]
@leonelgregioauler hi!!!

Thank you so much for simplifying the concept. I tried your implementation but it kept on throwing an error "writing failed"

**Reply by leonelgregioauler** [reply]
Hi @pixellet14  ! 

Do you have you project on GitHub ? If yes, can you give me the URL ? 

If not, I give you more suggestions:

- Send me an image that shows your code
- How are you generation the BLOB file ?
- You can debbug your app running using chrome://inspect/#devices

**Reply by pixellet14** [reply]
Hi @leonelgregioauler . thank you so much for extending your help. I don't have my project online because it's kind of confidential. however, below is that piece of code where I am trying to implement the plugin's method of writing a file in a custom folder 

```
var actionButton = document.querySelector('.croppinginitiation');
        actionButton.addEventListener('click', async function() {

                $("#ultraLoader").show();
                var canvas = cropper.getCroppedCanvas();
                var dataUrl = canvas.toDataURL('image/jpeg');

                // Convert dataUrl to base64 string without the prefix
                var base64Data = dataUrl.split(';base64,')[1];

                // Unique filename
                var fileName = 'TTM_Timeline_' + Date.now() + '.jpg';

                try {
                    var savedFilePath = await writeFile({
                        data: base64Data,
                        filename: fileName,
                        folder: "PictureCoupons"
                    });
                    $("#ultraLoader").hide();
                    otherOptions(savedFilePath);
                } catch (error) {
                    console.error('Write failed:', error);
                    $("#ultraLoader").hide();
                }
            });
            }

            // Convert a data URL to a Blob
            function dataURLToBlob(dataUrl) {
            var parts = dataUrl.split(';base64,');
            var contentType = parts[0].split(":")[1];
            var raw = window.atob(parts[1]);
            var rawLength = raw.length;
            var uInt8Array = new Uint8Array(rawLength);

            for (var i = 0; i < rawLength; ++i) {
            uInt8Array[i] = raw.charCodeAt(i);
            }

            return new Blob([uInt8Array], {type: contentType});
            }

            // Error callback
            function onError(error) {
            console.error('Error: ' + error);
            }
```
the error I am getting is Write failed: ReferenceError: writeFile is not defined

**Reply by leonelgregioauler** [reply]
Hello @pixellet14 !

You must do it by this way:

const result = await cordova.plugins.safMediastore.writeFile({ "data": base64Data, "filename": fileName, "subFolder": `CP 3001/${subFolderName}` })

PS.: See that you must not have to use the parameter "**folder**"

![image](https://github.com/apache/cordova/assets/12154804/3c3687ea-c241-47f8-9d41-ea3f61418e9e)

**Reply by pixellet14** [reply]
> Hello @pixellet14 !
> 
> You must do it by this way:
> 
> const result = await cordova.plugins.safMediastore.writeFile({ "data": base64Data, "filename": fileName, "subFolder": `CP 3001/${subFolderName}` })
> 
> PS.: See that you must not have to use the parameter "**folder**"
> 
> ![image](https://github.com/apache/cordova/assets/12154804/3c3687ea-c241-47f8-9d41-ea3f61418e9e)
> 

@leonelgregioauler hi!!

No I got it later.. but the thing is, it opens the file explorer and asks the user to save the file, while even providing the option to change the file name and it doesn't return the path the user chooses to save the file (upon saving). What if I want to then use the saved file path to upload it to the server for allowing the user to post it as a wall post or a status post or a chat message etc. Can you advise something here?

**Reply by Geshaa** [reply]
Hello @leonelgregioauler , I tried using your example but I have errors when I try to build android like these - https://prnt.sc/hdnyibDn6knQ 

This happens using Cordova 12, android 12.0.0 and this plugin - https://www.npmjs.com/package/cordova-plugin-saf-mediastore

Any ideas how I can resolve it?

**Reply by leonelgregioauler** [reply]
Hello @Geshaa !! Sorry for late !

There are a post bellow, but I put it again here !!

You need to edit the file build.gradle in the folder: **hybrid/platforms/android/app** and include the line 304 (see the image).
This line is used to import "DocumentFile" lib that doesn't exists.


![image](https://github.com/apache/cordova/assets/12154804/5f9d5ce5-b4bb-47c0-9d7e-9d7b8c53889d)

**Reply by Geshaa** [reply]
Ouh, that right @leonelgregioauler , it worked and now I have no errors on the build, thank you a lot.

**Comment by leonelgregioauler** [comment]
I made the implementation of plugin...

![image](https://github.com/apache/cordova/assets/12154804/5f663c83-9dd8-4444-961b-9470f5cd5339)

... and I got this error when doing the build. 

![image](https://github.com/apache/cordova/assets/12154804/6e1b936c-d8a6-410b-87a5-2bbc88a8b7ba)


This is because the class of line 21 doesn't exists

![image](https://github.com/apache/cordova/assets/12154804/cd9fc3d4-f1e8-4a61-93a6-70017cbd4e47)


**Am I doing the import of Java Class correctly ?** 

![image](https://github.com/apache/cordova/assets/12154804/8bec0ce4-8028-47e8-82c6-ef596044f81d)

**Reply by breautek** [reply]
It does seem their plugin is missing that dependency declaration.

But that's not how you import dependencies for your project.

There's another `build.gradle` file inside `app/` directory that should have a `dependencies` block and `implementation "androidx.documentfile:documentfile:1.0.1"` will need to be added.

The easiest way to do this is to fork their plugin (and perhaps make a PR) and add the following in their `plugin.xml` `android` platform block.

```xml
<framework src="androidx.documentfile:documentfile:1.0.1" />
```

Cordova will then add in the dependency to the project on plugin install.

**Reply by leonelgregioauler** [reply]
I made the import using the file build.gradle in the recommended folder and the import was done successfully. 

![image](https://github.com/apache/cordova/assets/12154804/c340c75a-4400-40e3-a1ef-f9acc6ecb56c)


Today I made the import in plugin.xml like you told me, but by this way, it doesn't work. Doesn't matter for now. I have been using the import by the other way (build.gradle)

About the implementation of this plugin ( "cordova-plugin-saf-mediastore"), **it works fine**. I had test almost all functions, but I had use only "writeFile" function in my project.

I think that the layout of the intents opened by "selectFolder", "selectFile", "openFolder" for example, are "bad".

I liked of the "openFile" but I haven't fully implemented yet, I'm still learning over this.

We can close this thread !!

Thank you so much.

**Reply by breautek** [reply]
Glad it worked...

For future readers:
<details>
<summary>~striked out bad advice~</summary>

```xml
<framework src="androidx.documentfile:documentfile:1.0.1" />
```

should have been:

```xml
<framework src="androidx.documentfile:documentfile" version="1.0.1" />
```
</details>

~I think (going off memory but the `src` attribute isn't suppose to include the version information). This approach would be a better way to go because the native project will be constructed properly when the platform is added.~

Update: `<framework src="androidx.documentfile:documentfile:1.0.1" />` is indeed the correct way to use bring in a gradle/android dependency.

If the contents of `platforms/android` is modified manually, them the changes will potentially be lost when adding or removing plugins, or when you upgrade the android platform (which involves removing it). If modifying platforms/android is required, then at least use a [hook](https://cordova.apache.org/docs/en/11.x/guide/appdev/hooks/index.html) to programmatically manipulate the native project so that the project generation is reproducible in a consistent manner.

The fix in the plugin.xml to include the `framework` directive would be ideal fix in this case however.

**Comment by rbaswawa** [comment]
Hi,

I have the same problem with cordova-plugin-saf-mediastore like leonelgregioauler had. 
I implemented both mentioned by breautek solutions. 
Unfortunately results in both cases are negative.

Solution with `<framework src="androidx.documentfile:documentfile" version="1.0.1" />` in plugin.xml  causes the same error: "..\SafMediastore.java:21:  package androidx.documentfile.provider does not exist".

Solution with `implementation "androidx.documentfile:documentfile:1.0.1"` in cordova app build.gradle fired error "Could not find method implementation() for arguments [androidx.documentfile:documentfile:1.0.1] on object of type org.gradle.api.internal.artifacts.dsl.dependencies.DefaultDependencyHandler."

My project is Quasar 2.12.6 app. I'm using cordova-android 12.0.1, java SDK 11, gradle 7.4.2.

Could you please give me some suggestions what im doing wrong and what should I do?

This cordova-plugin-saf-mediastore plugin seems something "fresh" and there are no many sources about it even in the universe of internet :-)

Thanks in advance and have a nice day
Andy

**Reply by breautek** [reply]
Based on the [docs](https://cordova.apache.org/docs/en/dev/plugin_ref/spec.html#framework)

`<framework src="androidx.documentfile:documentfile:1.0.1" />` is the correct approach, I was wrong before. Quite annoying cause more other directives has a `version` or `spec` attribute! :smile:  In fact I think I drew my example from existing code in my own project which was definitely not working properly. Just to make it that clear... 

> object of type org.gradle.api.internal.artifacts.dsl.dependencies.DefaultDependencyHandler."

This feels like it's saying that your configured repositories doesn't contain `androidx.documentfile:documentfile:1.0.1`

I don't know anything about Quasar but Cordova has this inside `platforms/android/app/repositories.gradle`:

```
ext.repos = {
    google()
    mavenCentral()
}
```

Is this file being overrided? If so it probably should still have `google()` and `mavenCentral()`

**Comment by rbaswawa** [comment]
Thank you very much for response.

I made check with `<framework src="androidx.documentfile:documentfile:1.0.1" />` and result is the same: package does not exist.
I also checked **platforms/android/app/repositories.gradle** file and it's content is exactly as you wrote.

I think Qusar has no negative impact on build but I can't be sure.
In order to build the project I run `quasar build -m android` but after while cordova do the job: 

> Running "cordova build --release android" in D:\ProgProjects\Quasar....\src-cordova"

and after some lines in terminal this appear:

>Task :app:compileReleaseJavaWithJavac FAILED
> D:\ProgProjects\Quasar..\src-cordova\platforms\android\app\src\main\java\com\customautosys\saf_mediastore\SafMediastore.java:21: error: package androidx.documentfile.provider does not exist
> import androidx.documentfile.provider.DocumentFile;.."

If you will have any idea what to do please let me know.
Thank you for your help.

Best regards
Andy

**Comment by leonelgregioauler** [comment]
Hello !

build.gradle exists in 2 folders. The correct file is in this path:

**hybrid/platforms/android/app**

Review this path, please !

**Comment by rbaswawa** [comment]
Hello leonelgregioauler

I can't find **hybrid/platforms/android/app** folder in my project.
I found build.gradle in: 
**src-cordova\platforms\android\build.gradle**
**src-cordova\platforms\android\app\build.gradle** (I put `implementation "androidx.documentfile:documentfile:1.0.1"` here)
**src-cordova\platforms\android\CordovaLib\build.gradle**

There are others build.gradle files in node_modules but I think they are not important (hope im right).
As you can see above I changed only one file (in accordance with breautek sugestions), others are original created by Quasar/cordova
Thank you for your help
Regards
Andy

**Reply by breautek** [reply]
To put knowledge here...

android/build.gradle is the "project" gradle and app/build.gradle & CordovaLib/build.gradle are "modules" for the application and Cordova framework respectively.

`app/build.gradle` is the proper place for app dependencies and is what the `<framework>` entries in `plugin.xml` would manipulate.

With that being said, I'm not sure on the cause of the original error:

> Could not find method implementation() for arguments [androidx.documentfile:documentfile:1.0.1] on object of type org.gradle.api.internal.artifacts.dsl.dependencies.DefaultDependencyHandler."

If you can open the native project in Android Studio, perhaps the IDE can give more details. (Just don't update AGP / Android Gradle Plugin)

**Reply by rbaswawa** [reply]
Hello!

I would like to confirm that solution with `implementation "androidx.documentfile:documentfile:1.0.1"` in app\build.gradle is correct in this case.

It seems that installing one smal plugin and make it working it is too much for one man.  I just put the code in wrong place inside build.gradle file and it was the reason why my build failed.

breautek and leonelgregioauler thank you very much for your help.
Im sorry disturbed you because of my mistake.

Regards
Andy
